dev/src/write-batch.ts

```diff
@@ -589,15 +575,198 @@ export class WriteBatch {
   }
 }
 
-/*!
+/**
+ * Validates the use of 'value' as a Precondition and enforces that 'exists'
+ * and 'lastUpdateTime' use valid types.
+ *
+ * @private
+ * @param arg The argument name or argument index (for varargs methods).
+ * @param value The object to validate
+ * @param allowExist Whether to allow the 'exists' preconditions.
+ */
+function validatePrecondition(
+    arg: string|number, value: unknown, allowExist: boolean): void {
+  if (typeof value !== 'object' || value === null) {
+    throw new Error('Input is not an object.');
+  }
+
+  const precondition = value as {[k: string]: unknown};
+
+  let conditions = 0;
+
+  if (precondition.exists !== undefined) {
+    ++conditions;
+    if (!allowExist) {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg, 'precondition')} "exists" is not an allowed condition.`);
+    }
+    if (typeof precondition.exists !== 'boolean') {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg, 'precondition')} "exists" is not a boolean.'`);
+    }
+  }
+
+  if (precondition.lastUpdateTime !== undefined) {
+    ++conditions;
+    if (!(precondition.lastUpdateTime instanceof Timestamp)) {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg,
+              'precondition')} "lastUpdateTime" is not a Firestore Timestamp.`);
+    }
+  }
+
+  if (conditions > 1) {
+    throw new Error(`${
+        invalidArgumentMessage(
+            arg, 'precondition')} Input contains more than one condition.`);
```

Comment:
This error and above use "condition" instead of "precondition". Is that intentional? We should probably use whatever leads to the most consistency across our APIs / docs / errors.